module/zfs/dsl_synctask.c

```diff
@@ -219,8 +221,10 @@ dsl_sync_task_sync(dsl_sync_task_t *dst, dmu_tx_t *tx)
 		/* MOS space is triple-dittoed, so we multiply by 3. */
 		if (used + dst->dst_space * 3 > quota) {
 			dst->dst_error = SET_ERROR(ENOSPC);
-			if (dst->dst_nowaiter)
+			if (dst->dst_freefunc != NULL) {
+				dst->dst_freefunc(dst->dst_arg);
 				kmem_free(dst, sizeof (*dst));
```

Comment:
Keeping this `kmem_free(dst, sizeof (*dst));` under the `dst->dst_freefunc != NULL` check makes passing a `freefunc` mandatory for a nowait caller.  Otherwise `dst` will be leaked here.  Keeping the `dst_nowaiter` would probably be the simplest way to handle this.